### ISSUE TITLE
New opam field to specify the location of the `.install` file

### ISSUE THREAD

**Issue by dbuenzli** [ocaml/opam#3880]
Currently a `.install` file needs to be written at the root of the package build directory to be considered. 

However the natural location of such a file is rather in the build directory of the build system: this make it easy to clean and need no special line in our `.gitignore`s. 

So I think it would be nice to have a field
```
install-file: <string>
```
that allows to specify its location w.r.t. to the root of the package build directory (the semantics of the `.install` file remains however exactly the same, i.e. relative file paths therein are expressed relative to the root of the package build directory). 

If something like https://github.com/ocaml/opam/issues/3797 becomes a reality end users won't be even bothered to have to write this down as most `opam` files will end-up being generated anyways.

**Comment by kit-ty-kate** [comment]
That's an interesting feature, i like that. It is not currently on my personal immediate todo-list (a fix for https://github.com/ocaml/opam/issues/5741 is my main priority for 2.5 and i expect this to take most of my time until the freeze), but someone else can come and contribute.

Adding a new `install-file` field before opam 3 is not possible because it would require a bump of the file format version (requiring to upgrade opam-repository) or for most people to use opam >= 2.3.0, see https://github.com/ocaml/opam/pull/5665.

We can get around that by introducing `x-install-file` instead and adding a lint check to make sure that the opam file has at least `available: opam-version >= "2.6.0"` (or whichever version that feature is added in). The main annoying part is that if someone is using a previous version of opam and the creator of the opam file forgot to add the `available: opam-version ...`, then there would be no way to tell the user that in fact nothing was installed. This feels a bit flimsy but it might not be too bad in practice.

**Comment by dbuenzli** [comment]
> Adding a new `install-file` field before opam 3 is not possible because it would

Yes I was actually thinking that versioning would be a issue.

I'm actually considering switching to non `.install` installs. But I wonder a bit about the performance cost. How does that work ? Do you diff the install prefix or stage the install in a clean directory and then copy over ? Or any other thing I'm not thinking about ?

**Comment by kit-ty-kate** [comment]
opam keeps the hash and metadata of every files installed in a switch and everytime an `install` field is executed, we have to scan the whole switch and figure out which files are new/changed either by their metadata (fast) or by their hash (slow). In practice it's reasonably fast on most unix system but i suspect it's fairly slow on Windows on large switches.

It is preferable to use an `.install` file whenever possible.

**Comment by dbuenzli** [comment]
I suspect the answer is no but is possible to specify multiple `.install` files. Basically I'd rather move [these instructions](https://erratique.ch/software/cmdliner/doc/cookbook.html#tip_tool_support) to the build step.

**Comment by kit-ty-kate** [comment]
It's sadly indeed not possible

**Comment by dbuenzli** [comment]
I'm tempted to patch the `.install` file written by the build system. They don't look hard to parse approximately in order to just add entries (famous last words). I'd rather not make `cmdliner` dependent on `opam-format` or `opam-installer` – in fact that may create a recursive dependency.

The idea is that in `build:` you invoke your build system, then add a line 
```
build:
[[ "build" ]
 [ "cmdliner" "install" "tool-support" "--update-opam-file={_:name}.install"
   "$(BUILD)/mytool" "_build/cmdliner-install` ]]
```

What do you think ?

**Comment by dbuenzli** [comment]
A bit OT but just for the record I [did that](https://github.com/dbuenzli/cmdliner/commit/d23d146f9f21f0f74b5194502a9e00ecf2d2a38e). This also has the advantage that it seem that in the `build:` section I don't run into [this bug](https://github.com/ocaml/opam/issues/6574).

**Comment by kit-ty-kate** [comment]
> What do you think ?

yeah that looks good to me

> I'd rather not make `cmdliner` dependent on `opam-format`

you could depend on the lower level `opam-file-format`. This at least makes sure that you get the whole syntax right

**Comment by dbuenzli** [comment]
> you could depend on the lower level [`opam-file-format`](https://github.com/ocaml/opam-file-format/).

In fact for installing `cmdliner` just depends on `ocaml` and `ocaml`'s build system (`make` as of writing). It's kind of a hold promise that this won't change so that people can use it to effortlessly devise excellent clis for their own bootstrapped systems. 

My [own collection](https://github.com/b0-system/) of build systems depend on `cmdliner`, they would be clearly offended in pulling in `dune` :–)
